File: ios/joinery-kit/Sources/JoineryMemberKit/ConversationModels.swift
import Foundation
import JoineryKit

/// One row of `conversation_list`.
public struct ConversationRow: Identifiable, Equatable, Sendable {
    public let conversationID: Int
    public let otherDisplayName: String
    public let preview: String
    public let lastMessageTime: String?
    public var unread: Bool
    public var muted: Bool
    public var id: Int { conversationID }

    init?(json: JSONValue) {
        guard let conversationID = json["conversation_id"]?.intValue else { return nil }
        self.conversationID = conversationID
        otherDisplayName = json["other_display_name"]?.stringValue ?? "Unknown"
        preview = json["preview"]?.stringValue ?? ""
        lastMessageTime = json["last_message_time"]?.stringValue
        unread = json["unread"]?.boolValue ?? false
        muted = json["muted"]?.boolValue ?? false
    }
}

/// The `conversation_list` payload. 20/page, matching the web inbox.
public struct ConversationPage: Equatable, Sendable {
    public static let perPage = 20

    public let conversations: [ConversationRow]
    public let totalCount: Int
    public let offset: Int
    public let perPage: Int

    public init?(data: JSONValue?) {
        guard let data else { return nil }
        conversations = (data["conversations"]?.arrayValue ?? []).compactMap(ConversationRow.init(json:))
        totalCount = data["total_count"]?.intValue ?? 0
        offset = data["offset"]?.intValue ?? 0
        perPage = data["per_page"]?.intValue ?? Self.perPage
    }
}

/// One message of `conversation_thread`.
public struct ThreadMessage: Identifiable, Equatable, Sendable {
    public let messageID: Int
    public let senderID: Int
    public let body: String
    public let time: String
    public let isMine: Bool
    public var id: Int { messageID }

    /// Build a message directly, e.g. to append the caller's own just-sent
    /// message locally (`conversation_send`'s response) without a round trip
    /// through JSON.
    public init(messageID: Int, senderID: Int, body: String, time: String, isMine: Bool) {
        self.messageID = messageID
        self.senderID = senderID
        self.body = body
        self.time = time
        self.isMine = isMine
    }

    init?(json: JSONValue) {
        guard let messageID = json["message_id"]?.intValue else { return nil }
        self.messageID = messageID
        senderID = json["sender_id"]?.intValue ?? 0
        body = json["body"]?.stringValue ?? ""
        time = json["time"]?.stringValue ?? ""
        isMine = json["is_mine"]?.boolValue ?? false
    }
}

/// The `conversation_thread` payload. `to` compose-mode dedup surfaces as
/// `isComposeMode` with a nil `conversationID` and no messages; the first
/// send creates the conversation and the send result carries its id.
public struct ThreadPayload: Equatable, Sendable {
    public let isComposeMode: Bool
    public let conversationID: Int?
    public let otherDisplayName: String
    public let otherUserID: Int?
    public let isMuted: Bool
    public let messages: [ThreadMessage]
    public let hasMore: Bool

    public init?(data: JSONValue?) {
        guard let data else { return nil }
        isComposeMode = data["is_compose_mode"]?.boolValue ?? false
        conversationID = data["conversation_id"]?.intValue
        otherDisplayName = data["other_display_name"]?.stringValue ?? "Unknown"
        otherUserID = data["other_user_id"]?.intValue
        isMuted = data["is_muted"]?.boolValue ?? false
        messages = (data["messages"]?.arrayValue ?? []).compactMap(ThreadMessage.init(json:))
        hasMore = data["has_more"]?.boolValue ?? false
    }
}

/// The `conversation_send` payload: the created message.
public struct SentMessage: Equatable, Sendable {
    public let conversationID: Int
    public let messageID: Int
    public let body: String
    public let sentTime: String

    public init?(data: JSONValue?) {
        guard let data, let messageID = data["message_id"]?.intValue else { return nil }
        self.messageID = messageID
        conversationID = data["conversation_id"]?.intValue ?? 0
        body = data["body"]?.stringValue ?? ""
        sentTime = data["sent_time"]?.stringValue ?? ""
    }
}

File: android/joinery-android-member/src/main/java/com/getjoinery/memberkit/ConversationModels.kt
package com.getjoinery.memberkit

import com.getjoinery.android.JsonValue

/** One row of `conversation_list`. */
data class ConversationRow(
    val conversationId: Int,
    val otherDisplayName: String,
    val preview: String,
    val lastMessageTime: String?,
    val unread: Boolean,
    val muted: Boolean,
) {
    companion object {
        fun from(json: JsonValue): ConversationRow? {
            val conversationId = json["conversation_id"]?.intValue ?: return null
            return ConversationRow(
                conversationId = conversationId,
                otherDisplayName = json["other_display_name"]?.stringValue ?: "Unknown",
                preview = json["preview"]?.stringValue ?: "",
                lastMessageTime = json["last_message_time"]?.takeUnless { it.isNull }?.stringValue,
                unread = json["unread"]?.boolValue ?: false,
                muted = json["muted"]?.boolValue ?: false,
            )
        }
    }
}

/** The `conversation_list` payload. 20/page, matching the web inbox. */
data class ConversationPage(
    val conversations: List<ConversationRow>,
    val totalCount: Int,
    val offset: Int,
    val perPage: Int,
) {
    companion object {
        const val PER_PAGE = 20

        fun from(data: JsonValue?): ConversationPage? {
            if (data == null) return null
            return ConversationPage(
                conversations = (data["conversations"]?.arrayValue ?: emptyList())
                    .mapNotNull { ConversationRow.from(it) },
                totalCount = data["total_count"]?.intValue ?: 0,
                offset = data["offset"]?.intValue ?: 0,
                perPage = data["per_page"]?.intValue ?: PER_PAGE,
            )
        }
    }
}

/** One message of `conversation_thread`. */
data class ThreadMessage(
    val messageId: Int,
    val senderId: Int,
    val body: String,
    val time: String,
    val isMine: Boolean,
) {
    companion object {
        fun from(json: JsonValue): ThreadMessage? {
            val messageId = json["message_id"]?.intValue ?: return null
            return ThreadMessage(
                messageId = messageId,
                senderId = json["sender_id"]?.intValue ?: 0,
                body = json["body"]?.stringValue ?: "",
                time = json["time"]?.stringValue ?: "",
                isMine = json["is_mine"]?.boolValue ?: false,
            )
        }
    }
}

/**
 * The `conversation_thread` payload. `to` compose-mode dedup surfaces as
 * `isComposeMode` with a null `conversationId` and no messages; the first send
 * creates the conversation and the send result carries its id.
 */
data class ThreadPayload(
    val isComposeMode: Boolean,
    val conversationId: Int?,
    val otherDisplayName: String,
    val otherUserId: Int?,
    val isMuted: Boolean,
    val messages: List<ThreadMessage>,
    val hasMore: Boolean,
) {
    companion object {
        fun from(data: JsonValue?): ThreadPayload? {
            if (data == null) return null
            return ThreadPayload(
                isComposeMode = data["is_compose_mode"]?.boolValue ?: false,
                conversationId = data["conversation_id"]?.takeUnless { it.isNull }?.intValue,
                otherDisplayName = data["other_display_name"]?.stringValue ?: "Unknown",
                otherUserId = data["other_user_id"]?.takeUnless { it.isNull }?.intValue,
                isMuted = data["is_muted"]?.boolValue ?: false,
                messages = (data["messages"]?.arrayValue ?: emptyList()).mapNotNull { ThreadMessage.from(it) },
                hasMore = data["has_more"]?.boolValue ?: false,
            )
        }
    }
}

/** The `conversation_send` payload: the created message. */
data class SentMessage(
    val conversationId: Int,
    val messageId: Int,
    val body: String,
    val sentTime: String,
) {
    companion object {
        fun from(data: JsonValue?): SentMessage? {
            if (data == null) return null
            val messageId = data["message_id"]?.intValue ?: return null
            return SentMessage(
                messageId = messageId,
                conversationId = data["conversation_id"]?.intValue ?: 0,
                body = data["body"]?.stringValue ?: "",
                sentTime = data["sent_time"]?.stringValue ?: "",
            )
        }
    }
}
